shared-bindings/ulab/__init__.c

```diff
@@ -0,0 +1,637 @@
+//|
+//| `ulab` is a numpy-like module for micropython, meant to simplify and
+//| speed up common mathematical operations on arrays. The primary goal was to
+//| implement a small subset of numpy that might be useful in the context of a
+//| microcontroller. This means low-level data processing of linear (array) and
+//| two-dimensional (matrix) data.
+//|
+//| `ulab` is adapted from micropython-ulab, and the original project's
+//| documentation can be found at
+//| https://micropython-ulab.readthedocs.io/en/latest/
+//|
+//| `ulab` is modeled after numpy, and aims to be a compatible subset where
+//| possible.  Numpy's documentation can be found at
+//| https://docs.scipy.org/doc/numpy/index.html
+//|
+//| .. contents::
+//|
+
+//| .. attribute:: __version__
+//|
+//| The closest corresponding version of micropython-ulab
+//|
+STATIC MP_DEFINE_STR_OBJ(ulab_version_obj, "0.26.7");
+
+//| ulab.array -- 1- and 2- dimensional array
+//| ==================================================
+//| .. class:: ulab.array(values, *, dtype=float)
+//|
+//|   :param sequence values: Sequence giving the initial content of the array.
+//|   :param dtype: The type of array values, ``int8``, ``uint8``, ``int16``, ``uint16``, or ``float``
+//|
+//|   The `values` sequence can either be a sequence of numbers (in which case a
+//|   1-dimensional array is created), or a sequence where each subsequence has
+//|   the same length (in which case a 2-dimensional array is created).
+//|
+//|   In many cases, it is more convenient to create an array from a function
+//|   like `zeros` or `linspace`.
+//|
+//|    .. method:: asbytearray()
+//|
+//|       Returns an `array.array` object that refers to the same memory.
+//|       This can be used with functions and methods that require "buffers",
+//|       such as `audiocore.RawSample`.
+//|
+//|       Despite the name, the byte array preserves the type information of
+//|       the array.
+//|
+//|       **WARNING:** `asbytearray` is a `ulab`-only method; it has no
+//|       equivalent in numpy.
+//|
+//|    .. method:: flatten(*, order='C')
+//|
+//|       :param order: Whether to flatten by rows ('C') or columns ('F')
+//|
+//|       Returns a new `ulab.array` object which is always 1 dimensional.
+//|       If order is 'C' (the default", then the data is ordered in rows;
+//|       If it is 'F', then the data is ordered in columns.  "C" and "F" refer
+//|       to the typical storage organization of the C and Fortran languages.
+//|
+//|    .. method:: rawsize()
+//|
+//|       Returns a 5-tuple with the following data
+//|
+//|       #. number of rows
+//|       #. number of columns
+//|       #. length of the storage (should be equal to the product of 1. and 2.)
+//|       #. length of the data storage in bytes
+//|       #. size of one element in bytes
+//|
+//|       **WARNING:** `rawsize` is a `ulab`-only method; it has no
+//|       equivalent in ``numpy``.
```

Comment:
How is this used? Can we remove it or replace it with something that takes in a ulab.array instead? I'd love to be a strict subset so we can use this interchangeably with a numpy array in CPython.

Reply:
I don't think that rawsize is useful as anything but a debug tool and wouldn't mind turning it off.

Reply:
Ok, that would be my vote. I think python does actually have a way to pass in an object and get its size back. I can't remember what though.

Reply:
While I think that `rawsize` is actually quite useful in an embedded environment, I am going to move all numpy-incompatible functions/methods into a separate module that you can easily switch off. I believe, that should satisfactorily settle the issue.

Reply:
It is always a great please to respond to my own comments, but I just wanted to relay a "never mind" to myself. I will remove `rawsize`, and insert the `itemsize`, and `size'. That should be a fitting replacement.